### DISCUSSION TITLE
Chainable firstValueFrom and lastValueFrom

### DISCUSSION THREAD

**Discussion by csisy** [ReactiveX/rxjs#7064] [Ideas / Feature request]
Hello,

Before `toPromise` was marked as deprecated in v7, the chaining calls were easy to read:

```
myObservable
  .toPromise()
  .then(...)
  .then(...)
```

It was especially nice if the observable was not stored in a variable (which is a pretty common case, e.g. Angular http requests that return an `Observable`).

Now, the `firstValueFrom` and `lastValueFrom` breaks the code flow, since the observable must be wrapped inside the function call. And even worse, if we need to use other parameters of these mentioned functions.

I totally agree with the deprecation of `toPromise`, however, I think that both the `firstValueFrom` and `lastValueFrom` should be part of the `Observable` interface, just like the `toPromise` is.

Compare the clunky wrapped code:

```
firstValueFrom(...)
  .then(...)
```

to the well-known, easier to format toPromise-like code:

```
myObservable
  .firstValueFrom(...)
  .then(...)
```

However, in this case the function name should be a bit different, maybe removing `from` is enough.

**Comment by voliva** [comment]
I don't fully understand the disadvantage of the current API...

```ts
myObservable
  .toPromise()
  .then(...)
  .then(...)

// vs 

firstValueFrom(myObservable)
  .then(...)
  .then(...)
```

For me, both of them are readable, although different. If anything, the current API is more concise on what actual value you want to get.... On the previous one if `myObservable` didn't complete then it would not resolve the promise.

**Reply by csisy** [reply]
Again, I'm not against the change. The existence of firstValueFrom is totally fine.

And yes, your example is readable, but if myObservable is already multiline, then it becomes very... strange.

Additionally, the whole API is based around chaining method calls (in the pipe function, for example), I don't understand why this is different.

Lastly, it would not hurt anyone to put a firstValue function inside the Observable which simply calls the global firstValueFrom passing this as an argument. I think most people would use the chained call. Also it's much easier to migrate the code from the toPromise.
